File: hooks/lib/osengineer-git-cmd.js
'use strict';

// osengineer-git-cmd.js — token-walk git command classifier.
// Determines whether a shell command invokes a specific git subcommand.
// Handles: bare `git commit`, `git -C path commit`, `VAR=x git commit`,
// `/usr/bin/git commit`. The naive `^git\s+commit` regex misses 3 of 4.
//
// Origin: ported verbatim from get-shit-done/hooks/lib/git-cmd.js
// (SHA 40a442b21f8b7a0df252efdf4b6ac4defd9d3a1f) per ADR-001, renamed.

const path = require('path');

const ARGUMENT_TAKING_FLAGS = new Set([
  '-C', '--git-dir', '--work-tree', '--namespace', '--super-prefix',
  '--exec-path', '--html-path', '--man-path', '--info-path', '--list-cmds',
]);

const BOOLEAN_FLAGS = new Set([
  '-p', '--paginate', '--no-pager',
  '--no-replace-objects', '--bare',
  '--literal-pathspecs', '--glob-pathspecs', '--noglob-pathspecs',
  '--icase-pathspecs', '--no-optional-locks',
  '-P', '--no-lazy-fetch',
  '--version', '--help',
]);

function tokenize(cmd) {
  const tokens = [];
  let i = 0;
  const len = cmd.length;
  while (i < len) {
    while (i < len && /\s/.test(cmd[i])) i++;
    if (i >= len) break;
    let token = '';
    while (i < len && !/\s/.test(cmd[i])) {
      if (cmd[i] === "'") {
        i++;
        while (i < len && cmd[i] !== "'") token += cmd[i++];
        if (i < len) i++;
      } else if (cmd[i] === '"') {
        i++;
        while (i < len && cmd[i] !== '"') token += cmd[i++];
        if (i < len) i++;
      } else {
        token += cmd[i++];
      }
    }
    if (token) tokens.push(token);
  }
  return tokens;
}

function isGitSubcommand(cmd, sub) {
  if (!cmd || !sub) return false;
  const tokens = tokenize(cmd);
  let i = 0;
  while (i < tokens.length && /^[A-Za-z_][A-Za-z0-9_]*=/.test(tokens[i])) i++;
  if (i >= tokens.length) return false;
  const gitToken = tokens[i++];
  if (path.basename(gitToken) !== 'git') return false;
  while (i < tokens.length) {
    const t = tokens[i];
    const eqIdx = t.indexOf('=');
    const flagName = eqIdx !== -1 ? t.slice(0, eqIdx) : t;
    if (ARGUMENT_TAKING_FLAGS.has(flagName)) {
      i += (eqIdx !== -1) ? 1 : 2;
      continue;
    }
    if (BOOLEAN_FLAGS.has(t)) { i++; continue; }
    break;
  }
  if (i >= tokens.length) return false;
  return tokens[i] === sub;
}

module.exports = { isGitSubcommand, tokenize };

File: hooks/lib/osengineer_git_cmd.py
#!/usr/bin/env python3
# osengineer_git_cmd.py — token-walk git command classifier.
# Determines whether a shell command invokes a specific git subcommand.
# Handles: bare `git commit`, `git -C path commit`, `VAR=x git commit`,
# `/usr/bin/git commit`.

import re
from pathlib import Path

ARGUMENT_TAKING_FLAGS = {
    "-C", "--git-dir", "--work-tree", "--namespace", "--super-prefix",
    "--exec-path", "--html-path", "--man-path", "--info-path", "--list-cmds",
}

BOOLEAN_FLAGS = {
    "-p", "--paginate", "--no-pager",
    "--no-replace-objects", "--bare",
    "--literal-pathspecs", "--glob-pathspecs", "--noglob-pathspecs",
    "--icase-pathspecs", "--no-optional-locks",
    "-P", "--no-lazy-fetch",
    "--version", "--help",
}


def tokenize(cmd):
    tokens = []
    i = 0
    length = len(cmd)
    while i < length:
        while i < length and cmd[i].isspace():
            i += 1
        if i >= length:
            break
        token = ""
        while i < length and not cmd[i].isspace():
            if cmd[i] == "'":
                i += 1
                while i < length and cmd[i] != "'":
                    token += cmd[i]
                    i += 1
                if i < length:
                    i += 1
            elif cmd[i] == '"':
                i += 1
                while i < length and cmd[i] != '"':
                    token += cmd[i]
                    i += 1
                if i < length:
                    i += 1
            else:
                token += cmd[i]
                i += 1
        if token:
            tokens.append(token)
    return tokens


def is_git_subcommand(cmd, sub):
    if not cmd or not sub:
        return False
    tokens = tokenize(cmd)
    i = 0
    while i < len(tokens) and re.match(r"^[A-Za-z_][A-Za-z0-9_]*=", tokens[i]):
        i += 1
    if i >= len(tokens):
        return False
    git_token = tokens[i]
    i += 1
    if Path(git_token).name != "git":
        return False
    while i < len(tokens):
        t = tokens[i]
        eq_idx = t.find("=")
        flag_name = t[:eq_idx] if eq_idx != -1 else t
        if flag_name in ARGUMENT_TAKING_FLAGS:
            i += 1 if eq_idx != -1 else 2
            continue
        if t in BOOLEAN_FLAGS:
            i += 1
            continue
        break
    if i >= len(tokens):
        return False
    return tokens[i] == sub


if __name__ == "__main__":
    import sys
    if len(sys.argv) >= 3:
        print(is_git_subcommand(sys.argv[1], sys.argv[2]))
